hop/robust_publisher.py

```diff
@@ -0,0 +1,633 @@
+        """
+        return self.data[key]
+
+    def pop_highest_priority(self):
+        """Fetch the item with the highest priority (lowest key) currently in the queue,
+            and removes it from the queue.
+
+        Returns: The (key, value) tuple for the highest priority item in the queue, or None if the
+                 queue is empty.
+        """
+        if len(self.data) == 0:
+            return None
+        key = heapq.heappop(self.priorities)
+        value = self.data[key]
+        del self.data[key]
+        return (key, value)
+
+    def remove(self, key):
+        """Remove an item from the queue, regardless of its location.
+
+        Raises: KeyError if the key does not exist.
+        """
+        del self.data[key]
+        idx = self.priorities.index(key)
+        del self.priorities[idx]
+        heapq.heapify(self.priorities)
+
+    def __delitem__(self, key):
+        self.remove(key)
+
+
+def _ensure_bytes_like(thing):
+    """Force an object which may be string-like to be bytes-like
+
+    Args:
+        thing: something which might be a string or might already be encoded as bytes.
+
+    Return:
+        Either the original input object or the encoding of that object as bytes.
+    """
+    try:  # check whether thing is bytes-like
+        memoryview(thing)
+        return thing  # keep as-is
+    except TypeError:
+        return thing.encode("utf-8")
+
+
+class PublicationJournal:
+    """
+    An object which tracks the state of messages which are being sent, persists that state to disk,
+    and enables it to be restored if the program stops unexpectedly.
+    """
+
+    int_format = "!Q"
+    int_size = struct.calcsize(int_format)
+    crc_format = "!I"
+    crc_size = struct.calcsize(crc_format)
+    msg_record_type = 0
+    sent_record_type = 1
+
+    @staticmethod
+    def encode_int(value: int):
+        return(struct.pack(PublicationJournal.int_format, value))
+
+    @staticmethod
+    def decode_int(buffer: bytes):
+        return struct.unpack(PublicationJournal.int_format, buffer)[0]
+
+    @staticmethod
+    def encode_crc(crc: int):
+        return(struct.pack(PublicationJournal.crc_format, crc))
+
+    @staticmethod
+    def decode_crc(buffer: bytes):
+        return struct.unpack(PublicationJournal.crc_format, buffer)[0]
+
+    def __init__(self, journal_path="publisher.journal"):
+        """Prepare a journal, including loading any data previously persisted to disk.
+
+        Args:
+            journal_path: The filesystem path from/to which the journal data should be read/written.
+
+        Raises:
+            PermissionError: If existing journal file does not have suitable permissions.
+            RuntimeError: If existing journal data cannot be read.
+        """
+        self.journal_path = journal_path
+        # each queue will store (message, headers) tuples, keyed by sequence numbers
+        self.messages_to_send = RAPriorityQueue()
```

Comment:
This seems less clear to me since it would then not mention the purpose of the queue (as there are two queues, currently `messages_to_send` and `messages_to_send`)?